Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift

```diff
@@ -0,0 +1,510 @@
+    ///            on the completion of the task.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    @discardableResult
+    @inlinable
+    func scheduleTask<T>(
+        deadline: NIODeadline,
+        _ task: @escaping () throws -> T
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.scheduleTask(deadline: deadline) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+
+    /// Schedule a `task` that is executed by this `EventLoop` after the given amount of time.
+    ///
+    /// - parameters:
+    ///     - task: The synchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the completion of the task.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    /// - note: The `in` value is clamped to a maximum when running on a Darwin-kernel.
+    @discardableResult
+    @inlinable
+    func scheduleTask<T>(
+        in delay: TimeAmount,
+        _ task: @escaping () throws -> T
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.scheduleTask(in: delay) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - Note: The `T` must be `Sendable` since the isolation domains of the event loop future returned from `task` and
+    /// this event loop might differ.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    @discardableResult
+    @inlinable
+    func flatScheduleTask<T: Sendable>(
+        deadline: NIODeadline,
+        file: StaticString = #file,
+        line: UInt = #line,
+        _ task: @escaping () throws -> EventLoopFuture<T>
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.flatScheduleTask(deadline: deadline, file: file, line: line) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+}
+extension EventLoop {
+    /// Assumes the calling context is isolated to the event loop.
+    @usableFromInline
+    func assumeIsolated() -> AssumeIsolatedEventLoop {
+        AssumeIsolatedEventLoop(self)
+    }
```

Comment:
I think the naming/behaviour here is a bit surprising. From reading the function name I would assume that _this_ function checks we are isolated to the EL, not the individual functions in `AssumeIsolatedEventLoop`. This is made more confusing by `AssumeIsolatedEventLoop` being `Sendable` too.

I know the pattern is the roughly the same as our sync views, however, the sync views predate `Sendable` (and aren't marked `Sendable`).  Perhaps we should change the pattern for the sync views such that they assert on EL when created and are explicitly not `Sendable`?

Reply:
I agree that those types shouldn't be `Sendable`. Pushed the changes up.